tlvf: CmduMessageRx: automatic tlv parse support

tomereli opened the pull request:
CmduMessageRx provides a `parse` API which as for now parses only the
CMDU header. Users are then expected to call `addClass` for each TLV manually.
This commit adds support for parse_tlvs option to the class's parse API,
disabled by default for backward compatability:
When parse_tlvs is true, the `parse` API goes through all the TLVs in
the CMDU and calls addClass for each found TLV.
If Unknown TLVs are encountered, addClass<unknownTlv> is called.
Parsing ends successfully only once tlvEndOfMessage is found and added.

This option is intended to be used only for standard CMDUs, such as the
ones defined in the MultiAP and 1905.1a specifications, since it is
using the TLV type for the casting.

Note that for WSC TLV, we support only M1 and M2 which are required by
the MultiAP specification. The two WSC types are differenciated via the
MSG_TYPE attribute. In the future, this will be also handled
automatically by the parser by adding each attribute manually without
using the tlvWscMx.yaml files which are wrong since they assume that the
attributes have to be the same order they are defined in the yaml.

Also add a test in tlvf_test which verifies the new functionality works. No effect on other components.

Commit:
tlvf: add templated version of getClass w/o index

CmduMessage GetClass API includes a templated version which accepts a
type to which it dynamically casts the returned shared_ptr to.
However, it requires the idx in the class vector, which is not always
available.
This commit adds another overload to the templated getClass API which
does not require the class index, therefore can be used to retrieve
parsed classes dynamically. If there are multiple classes of this type
in the class vector, then the first one is returned.

In addition, this commit modifies the templated version of getClass
which which accepts an index since it is not needed by current code.
The change is instead of returning the class at index `idx` in the
all-classes vector, the function now returns the class of the templated
type at index `idx` in the array of these type of classes.
For example, consider a CMDU which includes the following tlvs: [tlv1,
tlv2, tlv1].

- getClass<tlv1>() will return the first tlv - which is tlv1
- getClass<tlv1>(0) will return the first tlv of the tlv1 array
- getClass<tlv1>(1) will return the second tlv of the tlv1 array, which
is the third tlv in the all-tlvs array

Finally, a templated version of getClassCount API is added so that
users can iterate through all classes of a specific type as follows:

```
for (int i = 0; i < cmdu.getClassCount<tlv1>(); i++)
    auto c = cmdu.getClass<tlv1>(i);
```

The motivation for this is that once automatic parsing of standard CMDUs
is added to tlvf (in the followup commits), users will be able to
easilly get the class in which they are interested.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: CmduMessageRx: automatic tlv parse support

CmduMessageRx provides a `parse` API which as for now parses only the
CMDU header. Users are then expected to call `addClass` for each TLV manually.
This commit adds support for parse_tlvs option to the class's parse API,
disabled by default for backward compatability:
When parse_tlvs is true, the `parse` API goes through all the TLVs in
the CMDU and calls addClass for each found TLV.
If Unknown TLVs are encountered, addClass<unknownTlv> is called.
Parsing ends successfully only once tlvEndOfMessage is found and added.

This option is intended to be used only for standard CMDUs, such as the
ones defined in the MultiAP and 1905.1a specifications, since it is
using the TLV type for the casting.

Note that for WSC TLV, we support only M1 and M2 which are required by
the MultiAP specification. The two WSC types are differenciated via the
MSG_TYPE attribute. In the future, this will be also handled
automatically by the parser by adding each attribute manually without
using the tlvWscMx.yaml files which are wrong since they assume that the
attributes have to be the same order they are defined in the yaml.

This is a preperatory commit - it has no effect on functionality.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: test: add automatic tlv parser test

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

LiorAmram commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -31,12 +71,145 @@ CmduMessageRx::~CmduMessageRx()
+}
 
-std::shared_ptr<cCmduHeader> CmduMessageRx::parse(uint8_t *buff, size_t buff_len, bool swap_needed)
+std::shared_ptr<BaseClass> CmduMessageRx::parseNextTlv()
```
Did you manually write this method.........?

tomereli replied:
No, @Galvien updated tlvf in a different commit which generates the switch statement. I think however that since it is not bound to be changed too often, that we can have it here statically instead of generating it each time.

LiorAmram replied:
I think that's a bad practice to insert auto generated code into non-auto generated files.
A better approach would be exporting this logic to an external class (which would be autogenerated), and have this class as a member of CmduMessageRx. That way, "parseNextTlv" would only call this class' method.

tomereli replied:
Fixed in a big fixup splitting to new classes and files. We planned to do that anyway but since you asked so nicely.... :)

LiorAmram commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -46,7 +219,17 @@ std::shared_ptr<cCmduHeader> CmduMessageRx::parse(uint8_t *buff, size_t buff_len
+    if (!parse_tlvs)
+        return true;
+
+    while (auto tlv = parseNextTlv()) {
```
What does the standard say about additional TLVs after TLV EOM?
I think that if there are additional TLVs the frame is invalid, so it should be checked

tomereli replied:
The standard doesn't say anything about that, but having extraneous data after EOM TLV is fine and actually needed on some occasions, such as small packets (padding). We also saw that almost all golden units have extraneous data after EOM TLVs in some CMDUs, so we certainly shouldn't ignore such CMDUs.

Commit:
tlvf: Add getNextTlvData() API to CmduMessage

Add getNextTlvData() API which simillarly to getNextTlvType() and
getNextTlvLength() returns the current TLV's data pointer.
This will be used for manually parsing TLV data for automatic parsing
done in followup commits to parse WSC attributes, which are
currently not supported in TLVF (see #471)

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

morantr commented on framework/tlvf/AutoGenerated/include/tlvf/CmduMessage.h:
```diff
@@ -62,6 +62,15 @@ class CmduMessage {
+                return c;
+        }
+        return nullptr;
+    }
```
Am I missing something here? I don't understand the purpose of this loop.
The routine is looping through all the classes, from class idx=0 to the last class which was added, but return always on the first iteration (idx=0). So isn't it like always `return getClass<T>(0);`.

tomereli replied:
You are missing something probably, I'll try to explain:
`getClass<T>(idx)` is actually `std::dynamic_pointer_cast<T>(getClass(idx))`.
So, this basically goes through each class in the vector and tries to dynamically cast it to the requested type.

morantr replied:
Then why the `return? it will return in the first iteration, no?
Still, I don't see what benefit it gives to cast it.
Also, I didn't see any use if it.

tomereli replied:
> It will return in the first iteration, no?

No - `if (auto c = getClass<T>(idx))` will be `true` only if `getClass<T>(idx)` returned none NULL - which means the class at index `idx` is of type `T`.

> Also, I didn't see any use of it.

This is because this PR prepares all that's needed for automatic parsing but doesn't change any logic outside of TLVF to do the automatic parsing. It is used in `tlvf_test`:

```
CmduMessageRx received_message;
received_message.parse(recv_buffer, sizeof(recv_buffer), true, true);
auto tlv1_ = received_message.getClass<tlvNon1905neighborDeviceList>();
  if (!tlv1_)
    errors++;
```

As you can see - this function allows the user to simply call getClass<TYPE> and get nullptr if this TLV was not part of the CMDU, otherwise it returns the class it self already casted.

**Bottom line - this is one of the most important changes in this PR.**

morantr replied:
Thanks for the clarification! I can now understand that the loop will continue to iterate until the class in index `idx` will match to the function template T.
It wasn't so clear at the first look, so thanks.
Ok, but what you do if you have several TLVs from the same type?
For example Channel Selection Response message may contain One or **more** Channel Selection Response TLVs.
So I think you should always iterate from beginning to the end, and when you got a match, push it to std::list<std::shared_ptr<T>> which you will return as the function return value. If the list will be empty, then it means that the TLV was not found.

tomereli replied:
Most of the time you will just look for one class, so always returning a list is an overkill.
Moreover - The user of this API knows in advance if he has more than one element, in this case we will provide an additional API to return the class of type T at index idx.
We currently have such an API but it just tries to cast the class at idx `idx` out of *all classes* into type `T`. I will change it so that it will iterate through all classes, and only those which are of type `T` will be counted for - so it will behave as if there is an array which contains all the classes of type `T` and return the one at index `idx`. See last fixup for this.

morantr commented on framework/tlvf/AutoGenerated/include/tlvf/CmduMessageRx.h:
```diff
@@ -25,10 +23,12 @@ class CmduMessageRx : public CmduMessage {
+
+private:
+    std::shared_ptr<BaseClass> parseNextTlv();
+    std::shared_ptr<BaseClass> parseWscTlv();
```
I wonder if it correct to refer to a specific TLV in that class, and not give it a more general name like "parseAttributesTlv".

tomereli replied:
should be a nitpick - but note that this is initial implementation, which will be soon changed:
We will have a parent `CmduParser` class from which the following will derive:
- `TlvParser` which parses 1905 and mAP TLVs
- `WscAttrParser` which parses WSC attributes for WSC TLVs
- `IntelVsParser` which parses Intel Vendor specific messages

So name changes here are irrelevant, since again - they are bound to be changed soon.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -34,12 +73,174 @@ CmduMessageRx::~CmduMessageRx()
- */
+#pragma pack(push, 1)
+struct Tlv {
+    protected:
```
I wonder why did you choose protected and not private? Do you have plans to inherit it?
Same things on the "WSC".

tomereli replied:
Actually I took this from the transport code base, you are right - these can be private, or even public - we can define them as simple structs with functions. I'll change that in a fixup commit.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -34,12 +73,174 @@ CmduMessageRx::~CmduMessageRx()
+        uint8_t type() { return tlvType; }
+        Tlv *next() { return (Tlv *)((uint8_t *)this + size()); }
+        uint8_t *data() { return (uint8_t *)this + sizeof(Tlv); }
+        size_t size() { return sizeof(Tlv) + ntohs(tlvLength); }
```
You used here C-style casting. Since you can't use function notation casting on pointers, then the correct way to do it on CPP is with reinterpret_cast.
Same on WSC class, and parseWscTlv().

tomereli replied:
Accepted. Will change to `reinterpret_cast`.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -34,12 +73,174 @@ CmduMessageRx::~CmduMessageRx()
+
+// packed WSC Attribute struct to help in the parsing of the packet
+#pragma pack(push, 1)
+struct Wsc {
```
A more correct name would be "Attribute" IMO.

tomereli replied:
How about WscAttribute?

morantr replied:
The problem with WSC, it that it is specific to WSC. But theoretically, each tlv could have attributes like WSC. So I think a general name like "attribute" is more fit.

tomereli replied:
I've decided to go with `wscAttribute`. I don't like to depend on theory, better stick to what we know and have.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -34,12 +73,174 @@ CmduMessageRx::~CmduMessageRx()
+        uint16_t type() { return ntohs(wsc_type); }
+        Wsc *next() { return (Wsc *)((uint8_t *)this + size()); }
+        uint8_t *data() { return (uint8_t *)this + sizeof(Wsc); }
+        size_t size() { return sizeof(Wsc) + ntohs(length); }
```
Isn't `ntohs()` is required only if swap is needed?
Same thing on Tlv struct.

tomereli replied:
right - will fix. nice catch.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -34,12 +73,174 @@ CmduMessageRx::~CmduMessageRx()
+        }
+        case (1):{
+        return addClass<tlvAlMacAddressType>();
+        }
```
1. The clang-format.sh script would force zero indentation between the `switch` and `case`, and put an indentation between the `case` and the case body.
2. The brackets wrapping the tlv type values are redundant and can be removed.
3. the curly brackets can be removed as well because scoping is irrelevant here. It will squeeze the code length so I think it is a good idea.
4. I think it will be better to use the enum of the tlv type instead of an integer, even if it will be needed to create several switch cases for every TlvType enum (eTlvTypeMap/eTlvType). It will be more readable, and the reader will not need to guess of integer '#' is actually represent a specific tlv or not.
If not, at least represent it in hex notation is it will be easier to find.

tomereli replied:
This was autogenerated and in the future will be autogenerated during build, so @Galvien please take this into account and fix your script accordingly - see #467

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -34,12 +73,174 @@ CmduMessageRx::~CmduMessageRx()
+        case (16):{
+        return addClass<tlvSupportedFreqBand>();
+        }
+        #if 1
```
why `#if 1` ?

tomereli replied:
leftovers, will remove it.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -34,12 +73,174 @@ CmduMessageRx::~CmduMessageRx()
+    }
+}
+
+bool CmduMessageRx::parse(uint8_t *buff, size_t buff_len, bool swap_needed, bool parse_tlvs)
```
If you return a bool, you need to check if some places on the code still refer to the previous cmdu_header return value.
For example, on beerocks_message.cpp: print_cmdu_type();

tomereli replied:
Since it passed compilation I didn't bother (and since in most cases it is just used to check if the function call was successful - `if (!parse(...))`) - but i'll double check that I didn't miss other places.

tomereli replied:
I looked and its fine, cmdu_header is only used for checking if the function call was successful.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -49,7 +250,17 @@ std::shared_ptr<cCmduHeader> CmduMessageRx::parse(uint8_t *buff, size_t buff_len
+    if (!parse_tlvs)
+        return true;
+
+    while (auto tlv = parseNextTlv()) {
```
If there is an error and tlvEndofMessage is not exist, we could get stuck here on an infinite loop. I think you should add stop condition with a maximum number of iteration we would accept, and return false if we reach it.

tomereli replied:
Not needed, we will fail on parseNextTlv() which will reach the end of the buffer. No infinite loop here, error code will be handled internally.

morantr commented on framework/tlvf/AutoGenerated/src/CmduMessageRx.cpp:
```diff
@@ -49,7 +250,17 @@ std::shared_ptr<cCmduHeader> CmduMessageRx::parse(uint8_t *buff, size_t buff_len
+        return true;
+
+    while (auto tlv = parseNextTlv()) {
+        if (std::dynamic_pointer_cast<tlvEndOfMessage>(tlv)) {
```
I wonder if the dynamic pointer cast could return success if the TLV is not tlvEndOfMessage. I'm not sure.
I would explicitly compare the tlv type to the tlvEndOfMessage type.

tomereli replied:
This is exactly what is done in `parseNextTlv()`. It will fail if it is not exactly this class, so no need to do any explicit comparisons.

LiorAmram commented on framework/tlvf/src/src/CmduMessageRx.cpp:
```diff
@@ -7,6 +7,45 @@
  */
 
 #include <tlvf/CmduMessageRx.h>
+#include <tlvf/ieee_1905_1/tlvEndOfMessage.h>
```
Same here, all of these includes don't belong here.
Same external class solution I suggested would solve this as well.

tomereli replied:
Done

morantr commented on framework/tlvf/AutoGenerated/include/tlvf/CmduParser.h:
```diff
@@ -0,0 +1,34 @@
+class CmduParser {
+
+public:
+    CmduParser() = delete;
```
AFAIK There is no need to explicitly delete the empty constructor because another constructor is defined and it will be the default constructor. So even if you delete this line, this should get you a compilation error:
```CmduParser parser;```

tomereli replied:
There is no need - in this you are right, but the compilation error, in this case, is more readable and therefore I prefer to keep it. 
It is considered good coding practice to explicitly delete default constructor if it is not used.

morantr commented on framework/tlvf/AutoGenerated/src/CmduTlvParser.cpp:
```diff
@@ -0,0 +1,219 @@
+        {
+            return reinterpret_cast<wscAttr*>(reinterpret_cast<uint8_t*>(this) + size(swap));
+        }
+        uint8_t *data() { return reinterpret_cast<uint8_t*>(this) + sizeof(wscAttr); }
```
Although I don't agree with it, you defined that the `*` sign shall not be attached to the type.
So it should be `<uint8_t *> ` (with space between)
If the clang-format script would have run on this file, it would have to change it. Just saying... 
This comment apply to several places on this PR.

tomereli replied:
The whole of TLVF is not bound to clang format, so it serves little value to do that now.
